src/QsCompiler/CompilerOptimizations/Evaluation.fs

```diff
@@ -0,0 +1,366 @@
+
+    override this.onConditionalExpression (e1, e2, e3) =
+        let e1 = this.simplify e1
+        match e1.Expression with
+        | BoolLiteral a -> if a then (this.simplify e2).Expression else (this.simplify e3).Expression
+        | _ -> CONDITIONAL (e1, this.simplify e2, this.simplify e3)
+
+    override this.onEquality (lhs, rhs) =
+        let lhs, rhs = this.simplify (lhs, rhs)
+        match isLiteral callables lhs.Expression && isLiteral callables rhs.Expression with
+        | true -> BoolLiteral (lhs.Expression = rhs.Expression)
+        | false -> EQ (lhs, rhs)
+
+    override this.onInequality (lhs, rhs) =
+        let lhs, rhs = this.simplify (lhs, rhs)
+        match isLiteral callables lhs.Expression && isLiteral callables rhs.Expression with
+        | true -> BoolLiteral (lhs.Expression <> rhs.Expression)
+        | false -> NEQ (lhs, rhs)
+
+    override this.onLessThan (lhs, rhs) =
+        this.arithBoolBinaryOp LT (<) (<) (<) lhs rhs
+
+    override this.onLessThanOrEqual (lhs, rhs) =
+        this.arithBoolBinaryOp LTE (<=) (<=) (<=) lhs rhs
+
+    override this.onGreaterThan (lhs, rhs) =
+        this.arithBoolBinaryOp GT (>) (>) (>) lhs rhs
+
+    override this.onGreaterThanOrEqual (lhs, rhs) =
+        this.arithBoolBinaryOp GTE (>=) (>=) (>=) lhs rhs
+
+    override this.onLogicalAnd (lhs, rhs) =
+        let lhs = this.simplify lhs
+        match lhs.Expression with
+        | BoolLiteral true -> (this.simplify rhs).Expression
+        | BoolLiteral false -> BoolLiteral false
+        | _ -> AND (lhs, this.simplify rhs)
+
+    override this.onLogicalOr (lhs, rhs) =
+        let lhs = this.simplify lhs
+        match lhs.Expression with
+        | BoolLiteral true -> BoolLiteral true
+        | BoolLiteral false -> (this.simplify rhs).Expression
+        | _ -> OR (lhs, this.simplify rhs)
+
+    override this.onAddition (lhs, rhs) =
+        let lhs, rhs = this.simplify (lhs, rhs)
+        match lhs.Expression, rhs.Expression with
+        | BigIntLiteral a, BigIntLiteral b -> BigIntLiteral (a + b)
+        | DoubleLiteral a, DoubleLiteral b -> DoubleLiteral (a + b)
+        | IntLiteral a, IntLiteral b -> IntLiteral (a + b)
+        | ValueArray a, ValueArray b -> ValueArray (a.AddRange b)
+        | StringLiteral (a, a2), StringLiteral (b, b2) -> StringLiteral (NonNullable<_>.New (a.Value + b.Value), a2.AddRange b2)
+        | _ -> ADD (lhs, rhs)
+
+    override this.onSubtraction (lhs, rhs) =
+        this.arithNumBinaryOp SUB (-) (-) (-) lhs rhs
+
+    override this.onMultiplication (lhs, rhs) =
+        this.arithNumBinaryOp MUL (*) (*) (*) lhs rhs
+
+    override this.onDivision (lhs, rhs) =
+        this.arithNumBinaryOp DIV (/) (/) (/) lhs rhs
+
+    override this.onExponentiate (lhs, rhs) =
+        let lhs, rhs = this.simplify (lhs, rhs)
+        match lhs.Expression, rhs.Expression with
+        | BigIntLiteral a, IntLiteral b -> BigIntLiteral (BigInteger.Pow(a, Convert.ToInt32(b)))
```

Comment:
The silent conversion to int 32 still needs to be replaced - it will fail according to spec, but should not fail silently.

Reply:
I made it throw an ArgumentException if the exponent is too big for 32 bits